### DISCUSSION TITLE
Allow reference type fields to customize the label of the content choices

### DISCUSSION THREAD

**Discussion by markjaquith** [tinacms/tinacms#3311] [Ideas]
With a `list: true` object field, you can use `ui { itemProps() }` to set a `label` for each item, based on the values in that item.

But with a reference field, this is not supported, so your dropdown uses raw filenames, like this:

<img width="357" alt="CleanShot 2022-10-26 at 11 15 51@2x" src="https://user-images.githubusercontent.com/353790/198065924-561646f3-d90e-4f93-9083-f5d6417ff563.png">

This is not user-friendly. I want to be abstracting away the underlying file storage system from editors.

I want to be able to use `ui {itemProps() }` here as well, so I can grab the `title:` field from the underlying markdown and use that in the dropdown.

<img width="212" alt="CleanShot 2022-10-26 at 11 22 55@2x" src="https://user-images.githubusercontent.com/353790/198067657-a5587085-ef7e-4e07-81f4-b73206666279.png">

Thoughts?

**Comment by jeffsee55** [comment]
Thanks @markjaquith - we should open that up to customization but at the very least it should use the field marked as `isTitle: true` for the value here. Turning this into a ticket.

**Reply by markjaquith** [reply]
Agree: solid default that most will be happy with.

**Reply by Jasperrr91** [reply]
Customisation would be a great addition but that basic Title behaviour is kinda crucial so that would be much appreciated. Seeing that we currently only have a reference to the file name available, I do have a solution that sort of works for our editors but it ain't pretty:
```ts
  ui: {
    itemProps: (item) => {
      const label = item.team_member
        .split('/')
        .pop()
        .split('.')
        .shift()
        .split(/-/)
        .map((word) => word.charAt(0).toUpperCase() + word.slice(1))
        .join(' ');

      return {
        key: item.id,
        label,
      };
    },
  },
```

**Reply by appinteractive** [reply]
Hey, with the new default to fallback to titles, we now face the issue of an endless list of titles without context (in my case folders). We have a quite nested structure which is not reflected here.

<img width="953" alt="grafik" src="https://user-images.githubusercontent.com/432828/229631252-ada19f5e-7d65-408b-b53b-ef861d47dc3f.png">

Now I'm stuck with an endless list of titles, but no context. I would also need to be able to format (and even better, group) select list items.

Would it be possible to use itemProps also on selects? And even mybe add grouping to them by beeing able to "post process" the returned structure? 

Also a possibility would be provide a method as `items` so we could query and format the data on our own.

**Comment by appinteractive** [comment]
Just for reference (pun intended), I solved the issue for me by re-using parts of the Tinas `ReferenceSelect` and combing that with HeadlessUI which is shipped with Tina. My result is following, by showing the document title plus its (cleaned) id which is my slug / url.

<img width="790" alt="grafik" src="https://user-images.githubusercontent.com/432828/229842159-ed58696d-8116-4472-80ee-80909696e53a.png">

@jeffsee55 why is not something like that the default?

**Reply by logan-anderson** [reply]
@appinteractive This seems like the better default to me! Is your code open source? 


cc @spbyrne

**Reply by appinteractive** [reply]
Not currently, but I was thinking about a Tina Field Plugin Repository. I will try to bundle that one and maybe others (Grouped SEO/Open Graph Fields, etc.) into a new repo. Maybe someone also wants to contribute some components there. :)

**Reply by appinteractive** [reply]
What I would like to add still is filtering, as searching such a large list is quite painful.

**Reply by coreyaus** [reply]
A community plugin repository sounds like a great idea - we've got a few niche field plugins but also a couple that are fairly generic and could be useful to others (e.g. a ConditionalField that takes a `ui.displayIf` argument with a function that returns a boolean value to decide whether the field should be visible based on the current values in other fields)

Would your team be interested in starting a community plugins repo inside the `tinacms` GitHub org Logan?

**Reply by jeffsee55** [reply]
This would be great, it'd also be really helpful for us to see how the community is using some of the more nuanced APIs from Tina. We hope to greatly improve the field customization experience in the coming months so the more examples and use-cases, the better.

**Reply by scottgallant** [reply]
Hey @coreyaus and @appinteractive , you inspired us to create this repo (list of community plugins): https://github.com/tinacms/ecosystem/

If you have anything to add to it, we'd love that.

**Reply by appinteractive** [reply]
awesome @scottgallant I will deliver ;)

**Reply by Perni1984** [reply]
@appinteractive for Grouped SEO/Open Graph Fields etc. kindly have a look on https://github.com/pcode-at/tinacms-seo and let us know if you want to contribute - it's MIT licensed and we would be happy if we would get more feedback from the community on it.

**Comment by appinteractive** [comment]
In case someone has feedback, that's my working vision of a "good" reference field:

https://user-images.githubusercontent.com/432828/233649293-3f66a691-1d43-42b5-af94-85b0f9974842.mov

It also caches the results from the Tina API and displays them right away on the next time, so you don't have a waiting screen 

Would wish we had something like that for "Links" too.

**Reply by scottgallant** [reply]
@appinteractive very impressive. Do you have a working code example you can link to?

**Reply by appinteractive** [reply]
Thanks, I will try to publish something till mid of next week. Need to clean up a bit and generalize some parts.

**Reply by appinteractive** [reply]
@scottgallant #3898 prepares the select to get some customization approaches right before I add the reference field. Being thankful for any feedback. 

I tried to do that by an external library but I'm not a big fan of react (mostly using Vue) as I was not able to ship the tailwind styles with the component so I decided to add it directly.

**Reply by scottgallant** [reply]
@appinteractive this is very cool!  I think a lot of people in the community would also love to use a font-selector like the one you showed in your demo video.  Doe you have an isolated demo of that we could link to from [here](https://github.com/tinacms/awesome-tinacms) ?

**Reply by appinteractive** [reply]
@scottgallant I did implement it actually in the pool requests inside the iFrame demo app. But I could make it a independent package that uses the select from pr.

**Reply by scottgallant** [reply]
Oh, I see that now.  That would be a good independent package if you're feeling up for it. No worries, if not.

**Reply by appinteractive** [reply]
I had issues to pack the styles for the packages currently as the tailwind styles get scrapped an using it in dev admin they are mostly unstyled. Do you have a suggestion for that? I can hack that by pasting the needed classes as comments inside the Tina config file but that's not a solution.

**Reply by appinteractive** [reply]
Regarding plugins opened a discussion #3912 an a repo

**Reply by scottgallant** [reply]
I saw that, @appinteractive . The team is thinking on it. Thank you!
